File: src/main/scala/com/github/phisgr/gatling/generic/SessionCombiner.scala
package com.github.phisgr.gatling.generic

import com.typesafe.scalalogging.Logger
import io.gatling.core.session.Session

import scala.annotation.varargs
import scala.util.control.NonFatal

trait SessionCombiner {
  /**
   * Combines the main session and the branched session
   *
   * @return a new session from both the branched and main flow
   */
  def reconcile(main: Session, branched: Session): Session

  final def combineSafely(main: Session, branched: Session, logger: Logger): Session = try {
    reconcile(main = main, branched = branched)
  } catch {
    case NonFatal(e) =>
      logger.warn("Session combining failed", e)
      main.markAsFailed
  }
}


object SessionCombiner {
  val NoOp: SessionCombiner = (main: Session, _: Session) => main

  @varargs
  def pick(attributes: String*): SessionCombiner = (main: Session, branched: Session) => {
    attributes.foldLeft(main) { case (acc, key) =>
      branched.attributes.get(key) match {
        case Some(value) => acc.set(key, value)
        case None => acc
      }
    }
  }
}

File: gatling-kt-ext/src/main/kotlin/com/github/phisgr/gatling/kt/generic/SessionCombiner.kt
package com.github.phisgr.gatling.kt.generic

import io.gatling.javaapi.core.Session
import com.github.phisgr.gatling.generic.SessionCombiner as SessionCombinerS
import io.gatling.core.session.Session as SessionS

@FunctionalInterface
fun interface SessionCombiner : SessionCombinerS {
    companion object {
        @JvmField
        val NO_OP: SessionCombiner = SessionCombiner { main, _ -> main }

        @JvmStatic
        fun pick(vararg attributes: String): SessionCombiner = SessionCombiner { main, branched ->
            val unwrapped = main.asScala()
            val updated = attributes.fold(unwrapped) { acc, attributeName ->
                when (val value = branched.get<Any>(attributeName)) {
                    null -> acc
                    else -> acc.set(attributeName, value)
                }
            }
            Session(updated)
        }
    }

    override fun reconcile(main: SessionS, branched: SessionS): SessionS =
        reconcile(Session(main), Session(branched)).asScala()

    fun reconcile(main: Session, branched: Session): Session
}
